sparse_strips/vello_common/src/filter_effects.rs

```diff
@@ -0,0 +1,1124 @@
+// Copyright 2025 the Vello Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Filter effects API based on the W3C Filter Effects specification.
+//!
+//! This module provides a comprehensive filter system supporting both high-level
+//! CSS filter functions and low-level SVG filter primitives. The API is designed
+//! to follow the W3C Filter Effects Module Level 1 specification.
+//!
+//! See: <https://drafts.fxtf.org/filter-effects/>
+//!
+//! ## Implementation Status
+//!
+//! ### ✅ Implemented
+//!
+//! **Filter Functions:**
+//! - `Blur` - Gaussian blur effect
+//!
+//! **Filter Primitives (Single Use Only):**
+//! - `Flood` - Solid color fill
+//! - `GaussianBlur` - Gaussian blur filter
+//! - `DropShadow` - Drop shadow effect (compound primitive)
+//!
+//! **Note:** Currently only single primitive filters are supported. Filter graphs with
+//! multiple connected primitives are not yet implemented.
+//!
+//! ### 🚧 Not Yet Implemented
+//!
+//! **Core Features:**
+//! - `FilterGraph` execution - Chaining multiple filter primitives together
+//! - `FilterInputs` - Connecting primitives to create complex effects
+//!
+//! **Filter Functions:**
+//! - `Brightness`, `Contrast`, `Grayscale`, `HueRotate`, `Invert`,
+//!   `Opacity`, `Saturate`, `Sepia`
+//!
+//! **Filter Primitives:**
+//! - `ColorMatrix` - Matrix-based color transformation
+//! - `Offset` - Geometric offset/translation
+//! - `Composite` - Porter-Duff compositing operations
+//! - `Blend` - Blend mode operations
+//! - `Morphology` - Dilate/erode operations
+//! - `ConvolveMatrix` - Custom convolution kernels
+//! - `Turbulence` - Perlin noise generation
+//! - `DisplacementMap` - Pixel displacement
+//! - `ComponentTransfer` - Per-channel transfer functions
+//! - `Image` - External image reference
+//! - `Tile` - Tiling operation
+//! - `DiffuseLighting`, `SpecularLighting` - Lighting effects
+
+use crate::color::{AlphaColor, Srgb};
+use crate::kurbo::Rect;
+use alloc::collections::BTreeMap;
+use alloc::vec::Vec;
+#[cfg(not(feature = "std"))]
+use peniko::kurbo::common::FloatFuncs as _;
+
+/// The main filter system.
+///
+/// A filter combines a graph of filter primitives with optional spatial bounds.
+/// If bounds are specified, the filter only applies within that region.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Filter {
+    /// Filter graph defining the effect pipeline.
+    pub graph: FilterGraph,
```

Comment:
We might want to `Arc` `graph` because the current API necessitates much `Filter` cloning (which creates allocations due to duplicating the `Vec`/`BTreeMap` inside `FilterGraph`). 

Arc'ing here will reduce allocations while minimising API changes